core/feconf.py

```diff
@@ -859,6 +859,7 @@ def get_empty_ratings() -> Dict[str, int]:
 DASHBOARD_CREATE_MODE_URL = '%s?mode=create' % CREATOR_DASHBOARD_URL
 EDITOR_URL_PREFIX = '/create'
 EXPLORATION_DATA_PREFIX = '/createhandler/data'
+EXPLORATION_IMAGE_UPLOAD_PREFIX = '/createhandler/imageupload'
```

Comment:
I don't think we're expecting changes to `feconf.py` just for testing. You can add a constant in the test file like here:

https://github.com/oppia/oppia/blob/f74b606e8511cd4296b3c99aad37e53b66cca196/core/controllers/resources_test.py#L38

@aks681 do you agree with this assessment?

Reply:
I confirm that this is correct (though I'm not @aks681 :P)

But I'm also a bit confused about why this needs to be added here. Surely if that url endpoint is already exposed on the server, then it would already be in the codebase somewhere? So I don't get why a new URL endpoint needs to be defined either here or in the test.

Reply:
The `imageupload` endpoint is defined in `main.py`, but does not appear in `feconf.py`:

https://github.com/oppia/oppia/blob/f74b606e8511cd4296b3c99aad37e53b66cca196/main.py#L583

It looks like this endpoint is mostly used in the frontend (except for `_test.py` files), which might be why it never made it into `feconf.py`?

Reply:
Oh I see the issue. Actually my suggestion here is opposite to @U8NWXD -- it would be to change the main.py thing to a constant and add that constant to feconf.py, then use that constant everywhere (including in tests).

Here's the rationale: it boils down to the DRY principle ("don't repeat yourself"). If there are scattered "magic strings" all over the place that mean the same thing, there's nothing tying them together. If the endpoint string in main.py subsequently changes then everything becomes inconsistent.

Reply:
@U8NWXD what do you think / are you ok with this?

Reply:
@seanlip  @U8NWXD  I talked with Chris and he advised me to add a constant in feconf.py and then use that constant everywhere else in the codebase. I think this is what @seanlip also wants me to do. Correct me if I need to implement something else.
One more doubt, Is it the right approach to declare a string as a variable and then use its substrings everywhere? Like in this case, if we declare `'/createhandler/imageupload/<entity_type>/<entity_id>'` as a variable `upload_url` inside `main.py` and then we use `upload_url[0:27]` where ever it is required.

Reply:
No, don't do that. "27" is what is called a "magic number", and if you do things that way, other devs won't know what you are referring to.

In general, look at relevant places in the codebase to understand how things are done. I would suggest that you look at the main.py file and see how things are represented there, then follow a similar pattern. If you find yourself needing to use "magic numbers/constants", then something is probably off with the design.

Reply:
Ok @seanlip. By the way does the changes i have done looks good to you for now? Because later on, (if required) we can add or shift all the variables to `main.py` itself.

Reply:
In general the sort of thing I'm seeing in resources_test.py seems fine, yes. Just make sure that if main.py is using this constant too, the string there is replaced with the constant (basically there should be no instance of the raw '/imageupload' string anywhere in the codebase).

Reply:
Ok @seanlip. Although, I have checked all the files inside `core` folder and sub folders and replaced every instance of such a string with a constant. I will now do the same in the entire codebase including `main.py`.